File: data_viz/src/analysis/pages/FourierFreqs/patternScore.ts
/**
 * Consolidated periodic-pattern SCORE for the Fourier Frequencies page.
 *
 * The FFT amp spectrum says WHICH day freqs carry energy; recurrence
 * audits say which actually repeat. This module consolidates BOTH into
 * ONE bar per integer day freq d — the "periodic noticeable highs and
 * lows" score:
 *
 *   score(d) = (amp(d) / σ_band) × recEXT(d) × acfFrac(d)
 *
 *   • amp(d)/σ_band — NOTICEABILITY: the day's energy-merged FFT
 *     amplitude in units of the swing-band σ (σ_band² = Σ_{d'≤N/4}
 *     amp(d')² / 2). A cycle must stand out from the band's total
 *     swing energy before it is worth calling a pattern.
 *   • recEXT(d) — EXTREMA EVIDENCE: fraction of the window's
 *     prominence-filtered alternating swing highs/lows whose
 *     FULL-CYCLE period estimates (consecutive-gap sums, plus doubled
 *     single gaps as half-cycle corroboration) land within ±15% of d,
 *     normalized by the max possible cycles floor((N−d)/d) and capped
 *     at 1. Only extrema standing out by ≥ 1.5× the daily-change σ
 *     count — this is the "noticeable highs and lows" factor.
 *   • acfFrac(d) — ACF COHERENCE (the noise gate): fraction of the
 *     multiples m·d at which the MA-detrended window is significantly
 *     self-similar (biased acf(m·d) ≥ 1.96/√N, ≥ 1 full period of
 *     overlap). A recurring pattern is self-similar at EVERY multiple;
 *     noise at none — this factor alone takes pure-noise scores to ~0.
 *
 * Detrending: centered-MA residual (window L = odd(⌊N/4⌋)) removes
 * periods ≥ ~N/4 — the trend that would otherwise dilute both the ACF
 * and the extrema evidence. O(N·L), no FFT needed client-side.
 *
 * Auditable range: d ≤ N/3 — a period needs ≥ ~3 cycles in the window
 * before "always repeats" can be claimed; longer periods score 0 and
 * are flagged not auditable.
 *
 * Validation anchors (analyze/fourier_freqs/_study_consolidated.py):
 * a pure 10d sine scores ≈ 0.89 at 10d; 10d sine + strong trend + noise
 * still peaks at 10d; pure noise scores ≈ 0 everywhere; 000300's real
 * ~39-58d swing family reaches ~0.10 on a 750d window.
 */

/** Evidence-vs-day tolerance (a pool entry within ±15% of d is a hit). */
const TOL = 0.15;
/** Extrema prominence threshold in units of the daily-change σ. */
const K_PROM = 1.5;
/** 95% white-noise ACF confidence multiplier. */
const Z95 = 1.96;

/** Consolidated score audit for ONE integer day frequency. */
export interface PatternScoreAudit {
  /** Day period in trading days. */
  day: number;
  /** Consolidated score: ampNorm × recEXT × acfFrac (0 when not auditable). */
  score: number;
  /** amp(d) / σ_band — noticeability (can exceed 1; ≤ √2 for a pure tone). */
  ampNorm: number;
  /** Extrema-evidence fraction (pool hits / max possible cycles, cap 1). */
  recEXT: number;
  /** ACF coherence fraction (significant multiples / max possible). */
  acfFrac: number;
  /** Raw evidence-pool hits within ±15% of this day. */
  evidence: number;
  /** Max possible cycles within the window: max(1, floor((N−day)/day)). */
  maxRepeats: number;
  /** MEASURED repeats — multiples m with acf(m·day) ≥ 1.96/√N. */
  repeats: number;
  /** Mean acf across all multiples ∈ [−1, 1]. */
  avgAcf: number;
  /** False when day > N/3 — too few cycles to claim recurrence. */
  auditable: boolean;
}

/** Window-level results of the consolidated score audit. */
export interface PatternScoreResult {
  /** Per-day audits keyed by integer day freq (spectrum-merged days). */
  audits: Map<number, PatternScoreAudit>;
  /** Swing-band σ (yuan): sqrt(Σ_{d≤N/4} amp(d)² / 2). */
  sigmaBand: number;
  /** Number of prominence-filtered alternating extrema in the window. */
  nExtrema: number;
}

/** Centered-MA detrended residual — removes periods ≥ ~N/4 (the trend). */
function maResidual(w: readonly number[]): Float64Array {
  const N = w.length;
  const L = Math.max(3, Math.floor(N / 4) | 1); // odd window ~N/4
  const half = L >> 1;
  const cs = new Float64Array(N + 1); // prefix sums
  for (let i = 0; i < N; i++) cs[i + 1] = cs[i] + w[i];
  const r = new Float64Array(N);
  for (let i = 0; i < N; i++) {
    const lo = Math.max(0, i - half);
    const hi = Math.min(N, i + half + 1);
    r[i] = w[i] - (cs[hi] - cs[lo]) / (hi - lo);
  }
  return r;
}

/** Per-day ACF recurrence audit of a (detrended) series.
 *  Biased ACF; a multiple counts when acf(m·d) ≥ 1.96/√N. */
function acfRecurrence(
  x: Float64Array,
): Map<number, { repeats: number; maxRepeats: number; frac: number; avgAcf: number }> {
  const N = x.length;
  const out = new Map<number, { repeats: number; maxRepeats: number; frac: number; avgAcf: number }>();
  const maxDay = Math.floor(N / 2);
  let mean = 0;
  for (let i = 0; i < N; i++) mean += x[i];
  mean /= N;
  const xc = new Float64Array(N);
  let denom = 0;
  for (let i = 0; i < N; i++) {
    xc[i] = x[i] - mean;
    denom += xc[i] * xc[i];
  }
  const tau = Z95 / Math.sqrt(N);
  if (denom > 0) {
    const acf = new Float64Array(N);
    for (let lag = 0; lag < N; lag++) {
      let s = 0;
      for (let t = lag; t < N; t++) s += xc[t] * xc[t - lag];
      acf[lag] = s / denom;
    }
    for (let day = 2; day <= maxDay; day++) {
      const maxRepeats = Math.floor((N - day) / day);
      if (maxRepeats < 1) {
        out.set(day, { repeats: 0, maxRepeats: 0, frac: 0, avgAcf: 0 });
        continue;
      }
      let repeats = 0;
      let sum = 0;
      for (let m = 1; m <= maxRepeats; m++) {
        const a = acf[m * day];
        sum += a;
        if (a >= tau) repeats += 1;
      }
      out.set(day, { repeats, maxRepeats, frac: repeats / maxRepeats, avgAcf: sum / maxRepeats });
    }
  } else {
    for (let day = 2; day <= maxDay; day++) {
      out.set(day, { repeats: 0, maxRepeats: Math.floor((N - day) / day), frac: 0, avgAcf: 0 });
    }
  }
  return out;
}

/** Topographic prominence of w[i] as an extremum (sign +1 max, −1 min):
 *  height minus the higher of the two flanking interval minima, where
 *  each interval extends until the signal crosses the extremum's level
 *  (a higher peak / deeper valley) or the window border. */
function signedProminence(w: readonly number[], i: number, sign: 1 | -1): number {
  const N = w.length;
  const h = sign * w[i];
  let leftMin = h;
  for (let j = i - 1; j >= 0; j--) {
    const v = sign * w[j];
    if (v > h) break;
    if (v < leftMin) leftMin = v;
  }
  let rightMin = h;
  for (let j = i + 1; j < N; j++) {
    const v = sign * w[j];
    if (v > h) break;
    if (v < rightMin) rightMin = v;
  }
  return h - Math.max(leftMin, rightMin);
}

/** Alternating-extrema evidence pool of the window.
 *
 * Swing highs/lows must (a) be strict local extrema and (b) stand out
 * by ≥ K_PROM × σ(daily change). Consecutive kept extrema are forced
 * to alternate (first-of-run wins). The POOL of full-cycle period
 * estimates = consecutive-gap sums (hi→lo→hi) + doubled single gaps
 * (half-cycle corroboration). */
function extremaEvidencePool(w: readonly number[]): { pool: number[]; nExtrema: number } {
  const N = w.length;
  if (N < 3) return { pool: [], nExtrema: 0 };
  // σ of daily changes (population, ddof=0) — the noise floor
  const d = new Float64Array(N - 1);
  let dmean = 0;
  for (let i = 0; i < N - 1; i++) {
    d[i] = w[i + 1] - w[i];
    dmean += d[i];
  }
  dmean /= N - 1;
  let varc = 0;
  for (let i = 0; i < N - 1; i++) {
    const v = d[i] - dmean;
    varc += v * v;
  }
  const prom = K_PROM * Math.sqrt(varc / (N - 1));

  // strict local maxima / minima passing the prominence filter
  const ext: Array<{ idx: number; typ: 1 | -1 }> = [];
  for (let i = 1; i < N - 1; i++) {
    if (w[i] > w[i - 1] && w[i] > w[i + 1]) {
      if (signedProminence(w, i, 1) >= prom) ext.push({ idx: i, typ: 1 });
    } else if (w[i] < w[i - 1] && w[i] < w[i + 1]) {
      if (signedProminence(w, i, -1) >= prom) ext.push({ idx: i, typ: -1 });
    }
  }
  ext.sort((a, b) => a.idx - b.idx);
  // force alternation — keep the first of each same-type run
  const kept: Array<{ idx: number; typ: 1 | -1 }> = [];
  for (const e of ext) {
    if (kept.length === 0 || kept[kept.length - 1].typ !== e.typ) kept.push(e);
  }
  // half-cycle gaps between consecutive alternating extrema
  const gaps: number[] = [];
  for (let i = 1; i < kept.length; i++) gaps.push(kept[i].idx - kept[i - 1].idx);
  // full-cycle period estimates (empty when fewer than 2 gaps)
  const pool: number[] = [];
  if (gaps.length > 1) {
    for (let i = 1; i < gaps.length; i++) pool.push(gaps[i - 1] + gaps[i]);
    for (const g of gaps) pool.push(2 * g);
  }
  return { pool, nExtrema: kept.length };
}

/** Consolidated periodic-pattern score audit.
 *
 * @param closes Window close prices (chronological, ending at the
 *  spectrum's last_date). Drives all time-domain factors.
 * @param dayAmps Energy-merged FFT amplitude per integer day freq
 *  (day → amp, yuan) — the Fourier reference, already merged by the
 *  chart builder.
 * @param rangeDays The spectrum window length N (bin count basis).
 */
export function auditPatternScores(
  closes: readonly number[],
  dayAmps: ReadonlyMap<number, number>,
  rangeDays: number,
): PatternScoreResult {
  const N = closes.length;

  // σ_band — total swing energy of the band (periods ≤ N/4)
  let sb2 = 0;
  const bandMaxDay = Math.floor(rangeDays / 4);
  for (const [day, a] of dayAmps) {
    if (day <= bandMaxDay) sb2 += a * a;
  }
  const sigmaBand = Math.sqrt(sb2 / 2);

  const audits = new Map<number, PatternScoreAudit>();
  if (N < 8) return { audits, sigmaBand, nExtrema: 0 };

  // time-domain factors over d = 2..⌊N/2⌋
  const { pool, nExtrema } = extremaEvidencePool(closes);
  const acf = acfRecurrence(maResidual(closes));

  for (const day of dayAmps.keys()) {
    const amp = dayAmps.get(day) ?? 0;
    const ampNorm = sigmaBand > 0 ? amp / sigmaBand : 0;
    const auditable = day <= Math.floor(N / 3);
    if (day < 2 || day > Math.floor(N / 2)) {
      // beyond the auditable recurrence range (incl. the k=1 full-window
      // bin) — amp shows, but no recurrence claim is possible
      audits.set(day, {
        day, score: 0, ampNorm, recEXT: 0, acfFrac: 0,
        evidence: 0, maxRepeats: 0, repeats: 0, avgAcf: 0,
        auditable: false,
      });
      continue;
    }
    // extrema evidence: pool hits within ±TOL·day, normalized by the
    // max possible cycles, capped at 1
    const tol = TOL * day;
    let evidence = 0;
    for (const p of pool) {
      if (Math.abs(p - day) <= tol) evidence += 1;
    }
    const maxRepeats = Math.max(1, Math.floor((N - day) / day));
    const recEXT = Math.min(evidence / maxRepeats, 1);
    const a = acf.get(day) ?? { repeats: 0, maxRepeats, frac: 0, avgAcf: 0 };
    const score = auditable ? ampNorm * recEXT * a.frac : 0;
    audits.set(day, {
      day, score, ampNorm, recEXT, acfFrac: a.frac,
      evidence, maxRepeats, repeats: a.repeats, avgAcf: a.avgAcf,
      auditable,
    });
  }
  return { audits, sigmaBand, nExtrema };
}

File: analyze/recurring_cycles/pattern_score.py
"""Per-day recurring rise/drop periodicity factors — the SEMANTIC CORE.

For each sliding window of N close prices and each integer day period d
(2..N//2; array index j = d − 2) the audit produces THREE day-aligned
factors:

  amp(d)      — energy-merged FFT amplitude of the day: sqrt(Σ amp_k²)
                over the bins k whose period N/k rounds to d (half-UP
                rounding). The Fourier REFERENCE — how large the swings
                at that period are (yuan). NOT recurrence evidence by
                itself.

  count(d)    = recEXT(d) × acfFrac(d)               — the COUNT factor
                recEXT  — prominence-filtered alternating-extrema
                          evidence: pool hits within ±15% of d, over the
                          max possible cycles floor((N−d)/d), capped 1
                acfFrac — MA-detrended ACF recurrence: fraction of the
                          multiples m·d with biased acf ≥ 1.96/√N
                This is the RECURRENCE evidence: did price actually
                rise-and-drop repeatedly with spacing ≈ d. A one-off
                swing, a trend, or noise all score ~0.

  strength(d) = (amp(d)/σ_band) × count(d)           — the summarized
                recurring strength. amp(d) normalized by the swing-band
                σ (sqrt(Σ_{d′≤N/4} amp(d′)² / 2)) then gated by the
                recurrence count. 0 where not auditable (d > N/3 — under
                3 cycles in the window).

The headline period_days = argmax of strength — the period at which
price BOTH cycled repeatedly AND with meaningful swing amplitude.

GPU note: the ACF is computed via FFT (Wiener–Khinchin) so the whole
batch is vectorizable and reuses the shared cudf→cupy→cpu FFT routing
in _fft.py (cuDF implements no FFT). Extrema extraction uses
scipy.signal (C-implemented) per window — the only per-row loop.

PROXY NOTE: everything in this module is pure host numpy/scipy math on
ndarrays — NO cudf objects are touched. The inputs MUST be RAW host
numpy arrays (compute.py unwraps cudf.pandas proxy arrays at the
pandas→numpy boundary via _host_array): proxy-subclass arrays would
route every numpy op here through the cudf fast/slow dispatcher
(profiled at 128 of 136 s per long-history code — each small op sent
to cupy + a `.get()` device sync).
"""
from __future__ import annotations

import numpy as np
from scipy.signal import peak_prominences

from analyze.recurring_cycles._fft import _irfft, _rfft

# Audit constants.
TOL = 0.15    # evidence-vs-day tolerance (±15% of the day)
K_PROM = 1.5  # extrema prominence threshold in units of σ(daily change)
Z95 = 1.96    # 95% white-noise ACF confidence multiplier

_EMPTY_F64 = np.empty(0, dtype=np.float64)

# Window-chunk size for the batched ACF/audit — bounds working memory
# (2N-padded complex FFT block ≈ chunk × 2N × 16 B) per the project's
# GPU batch-sending rule.
_CHUNK = 2048


def _ma_residual(w: np.ndarray) -> np.ndarray:
    """Centered-MA residual of each row — removes periods ≥ ~N/4.

    L = max(3, floor(N/4) made odd).
    """
    Wc, N = w.shape
    L = N // 4
    if L % 2 == 0:
        L += 1
    L = max(3, L)
    half = L // 2
    cs = np.zeros((Wc, N + 1), dtype=np.float64)
    cs[:, 1:] = np.cumsum(w, axis=1)
    i = np.arange(N)
    lo = np.maximum(0, i - half)
    hi = np.minimum(N, i + half + 1)
    ma = (cs[:, hi] - cs[:, lo]) / (hi - lo)[None, :]
    return w - ma


def _acf_fft(x: np.ndarray) -> np.ndarray:
    """Biased ACF (lags 0..N−1) of each mean-subtracted row via FFT.

    Linear (non-circular) autocorrelation: zero-pad to 2N, rfft → |F|²
    → irfft. Denominator is the full-window Σx² of the row (biased
    estimator). Rows with zero energy return all-zero ACF.
    """
    Wc, N = x.shape
    denom = np.einsum("ij,ij->i", x, x)
    xpad = np.zeros((Wc, 2 * N), dtype=np.float64)
    xpad[:, :N] = x
    F = _rfft(xpad, axis=1)
    ac = _irfft(F * np.conj(F), n=2 * N, axis=1)[:, :N]
    out = np.zeros((Wc, N), dtype=np.float64)
    nz = denom > 0
    out[nz] = ac[nz] / denom[nz, None]
    return out


def _acf_frac(acf: np.ndarray, N: int, days: np.ndarray) -> np.ndarray:
    """Fraction of significant multiples per day — vectorized over rows.

    For each day d, maxRepeats = floor((N−d)/d) (≥ 1 for d ≤ N//2) and
    frac(d) = #{m ≤ maxRepeats : acf[m·d] ≥ 1.96/√N} / maxRepeats. The
    (day, m) lag indices are flattened once and grouped with
    np.add.reduceat — no Python loop over days.
    """
    tau = Z95 / np.sqrt(N)
    mmax = (N - days) // days  # ≥ 1 for all days ≤ N//2
    starts = np.concatenate([[0], np.cumsum(mmax)[:-1]]).astype(np.int64)
    total = int(mmax.sum())
    day_idx = np.repeat(np.arange(len(days)), mmax)
    m_arr = np.arange(total, dtype=np.int64) - np.repeat(starts, mmax) + 1
    lags = m_arr * days[day_idx]
    # int32, NOT bool — np.add on bools acts as logical OR, not a sum.
    hits = (acf[:, lags] >= tau).astype(np.int32)  # (Wc, total)
    return np.add.reduceat(hits, starts, axis=1) / mmax[None, :]


def _extrema_pools(w: np.ndarray) -> list[np.ndarray]:
    """Per-row extrema evidence pools of the RAW closes.

    STRICT local maxima / minima (candidate detection vectorized over
    the whole window matrix) passing a topographic prominence >=
    K_PROM × sigma(daily changes, ddof=0) — scipy.signal.peak_prominences
    computes the same signed prominence the former TS port implemented by
    hand. Kept extrema are forced to alternate, then the pool of
    full-cycle period estimates = consecutive-gap sums (hi->lo->hi) plus
    doubled single gaps (half-cycle corroboration).
    """
    Wc, N = w.shape
    sds = np.std(w[:, 1:] - w[:, :-1], axis=1, ddof=0)
    mid = w[:, 1:-1]
    mask_hi = (mid > w[:, :-2]) & (mid > w[:, 2:])  # strict local maxima
    mask_lo = (mid < w[:, :-2]) & (mid < w[:, 2:])  # strict local minima
    pools: list[np.ndarray] = []
    for r in range(Wc):
        row = w[r]
        thr = K_PROM * sds[r]
        hi_idx = np.flatnonzero(mask_hi[r]) + 1
        lo_idx = np.flatnonzero(mask_lo[r]) + 1
        prom_hi = peak_prominences(row, hi_idx)[0] if hi_idx.size else _EMPTY_F64
        prom_lo = peak_prominences(-row, lo_idx)[0] if lo_idx.size else _EMPTY_F64
        idx = np.concatenate([hi_idx[prom_hi >= thr], lo_idx[prom_lo >= thr]])
        typ = np.concatenate(
            [np.ones(int((prom_hi >= thr).sum()), dtype=np.int8),
             -np.ones(int((prom_lo >= thr).sum()), dtype=np.int8)]
        )
        order = np.argsort(idx, kind="stable")
        idx, typ = idx[order], typ[order]
        if len(idx) > 1:
            keep = np.concatenate([[True], np.diff(typ) != 0])
            idx = idx[keep]
        if len(idx) < 3:
            pools.append(_EMPTY_F64)
            continue
        gaps = np.diff(idx).astype(np.float64)
        pools.append(np.concatenate([gaps[:-1] + gaps[1:], 2.0 * gaps]))
    return pools


def _rec_ext(pools: list[np.ndarray], N: int, days: np.ndarray) -> np.ndarray:
    """Extrema-evidence fraction per day — pool hits / max cycles, cap 1.

    Vectorized interval formulation: pool value p hits day d iff
    |p − d| ≤ TOL·d  ⇔  d ∈ [p/1.15, p/0.85] — a CONTIGUOUS integer-day
    range. Every pool element contributes +1 across its range; per-row
    day hits are the cumsum of those range end-points, scattered for ALL
    rows at once via a single weighted bincount over flattened
    (row, day) indices — no per-row Python loop and no (pool × days)
    broadcast. (12× faster than the broadcast form; verified equivalent
    up to knife-edge float boundaries, where the interval form is the
    mathematically exact one.)
    """
    n_rows = len(pools)
    n_days = len(days)
    maxrep = np.maximum(1, (N - days) // days).astype(np.float64)
    d_min, d_max = int(days[0]), int(days[-1])

    lens = np.fromiter((p.size for p in pools), dtype=np.int64, count=n_rows)
    if int(lens.sum()) == 0:
        return np.zeros((n_rows, n_days), dtype=np.float64)

    pool_all = np.concatenate(pools)
    rows = np.repeat(np.arange(n_rows, dtype=np.int64), lens)

    d_lo = np.ceil(pool_all / (1.0 + TOL)).astype(np.int64)
    d_hi = np.floor(pool_all / (1.0 - TOL)).astype(np.int64)
    keep = (d_hi >= d_min) & (d_lo <= d_max)
    if not keep.any():
        return np.zeros((n_rows, n_days), dtype=np.float64)
    d_lo, d_hi, rows = d_lo[keep], d_hi[keep], rows[keep]
    np.clip(d_lo, d_min, d_max, out=d_lo)
    np.clip(d_hi, d_min, d_max, out=d_hi)

    # Scatter +1 at each range start and −1 after each range end on the
    # (row, day+1) grid; cumsum along the day axis reconstructs hits.
    width = n_days + 1
    flat = rows * width
    n_keep = len(rows)
    delta = np.bincount(
        np.concatenate([flat + (d_lo - d_min), flat + (d_hi - d_min + 1)]),
        weights=np.concatenate([np.ones(n_keep), -np.ones(n_keep)]),
        minlength=n_rows * width,
    ).astype(np.float64).reshape(n_rows, width)
    hits = np.cumsum(delta, axis=1)[:, :n_days]
    return np.minimum(hits / maxrep[None, :], 1.0)


def compute_pattern_scores(
    windows: np.ndarray,
    amplitudes: np.ndarray,
    range_days: int,
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Compute the per-day amp / count / strength blocks for one (code, N).

    Args:
        windows: (n_windows, range_days) RAW close-price windows (the
            same sliding windows the FFT used, before detrending).
        amplitudes: (n_windows, range_days // 2) one-sided bin amplitude
            spectrum per window (excl. DC) — the FFT block from compute.
        range_days: the window size N.

    Returns:
        (amp_block, count_block, strength_block), each (n_windows,
        N//2 − 1) float64 DAY-aligned: element j is integer day period
        d = j + 2. amp_block = energy-merged FFT amplitude per day
        (yuan); count_block = recEXT × acfFrac (0 for d > N//2, and the
        day axis already ends at N//2); strength_block =
        (amp/σ_band) × count, 0 for d > N//3 (not auditable — under 3
        cycles). All zeros when N < 8 (window too short to audit).
    """
    n_windows, n_bins = amplitudes.shape
    N = range_days
    n_days = max(0, N // 2 - 1)
    amp_block = np.zeros((n_windows, n_days), dtype=np.float64)
    count_block = np.zeros((n_windows, n_days), dtype=np.float64)
    strength_block = np.zeros((n_windows, n_days), dtype=np.float64)
    if n_windows == 0 or n_days == 0 or N < 8:
        return amp_block, count_block, strength_block

    # Raw host numpy math only (see module docstring PROXY NOTE) — the
    # blocks are filled in place by the split-out impl below.
    _fill_pattern_scores(
        windows, amplitudes, N, amp_block, count_block, strength_block,
    )
    return amp_block, count_block, strength_block


def _fill_pattern_scores(
    windows: np.ndarray,
    amplitudes: np.ndarray,
    N: int,
    amp_block: np.ndarray,
    count_block: np.ndarray,
    strength_block: np.ndarray,
) -> None:
    """Fill the three day-aligned blocks in place for one (code, N)."""
    n_windows, n_bins = amplitudes.shape
    n_days = amp_block.shape[1]

    # Day period of each bin: floor(N/k + 0.5) — half-UP rounding.
    ks = np.arange(1, n_bins + 1, dtype=np.float64)
    day_of_bin = np.floor(N / ks + 0.5).astype(np.int64)
    days = np.arange(2, N // 2 + 1, dtype=np.int64)  # day axis

    # One-hot day-merge matrix (n_bins × n_days): column j accumulates
    # Σ amp_k² over the bins rounding to day j — merged amp² per day via
    # ONE matmul. Bins whose day falls outside 2..N//2 (day N from k=1)
    # contribute nothing.
    day_pos = np.where(
        (day_of_bin >= 2) & (day_of_bin <= N // 2), day_of_bin - 2, -1
    )
    merge = np.zeros((n_bins, n_days), dtype=np.float64)
    valid = day_pos >= 0
    merge[valid, day_pos[valid]] = 1.0

    # σ_band per row: sqrt(Σ amp² over days ≤ N//4 / 2).
    band_mask = (days <= N // 4).astype(np.float64)
    auditable_day = (days <= N // 3).astype(np.float64)

    for lo in range(0, n_windows, _CHUNK):
        hi = min(lo + _CHUNK, n_windows)
        wc = windows[lo:hi]
        amp = amplitudes[lo:hi]

        # ---- time-domain count factors --------------------------------
        resid = _ma_residual(wc)
        resid = resid - resid.mean(axis=1, keepdims=True)
        acf = _acf_fft(resid)
        afrac = _acf_frac(acf, N, days)
        rext = _rec_ext(_extrema_pools(wc), N, days)
        count_day = rext * afrac  # (Wc, n_days)

        # ---- amp factors from the FFT block ---------------------------
        amp_sq = amp * amp
        merged_sq = amp_sq @ merge            # (Wc, n_days) merged amp²
        amp_day = np.sqrt(merged_sq)          # energy-merged amplitude
        sig_band = np.sqrt((merged_sq @ band_mask[:, None])[:, 0] / 2.0)
        safe_sig = np.where(sig_band > 0, sig_band, 1.0)
        amp_norm = amp_day / safe_sig[:, None]
        amp_norm[sig_band <= 0, :] = 0.0

        strength_day = amp_norm * count_day * auditable_day[None, :]

        amp_block[lo:hi] = amp_day
        count_block[lo:hi] = count_day
        strength_block[lo:hi] = strength_day
